tests/unittests/detection/test_map.py

```diff
@@ -660,3 +679,34 @@ def test_error_on_wrong_input():
             [{"boxes": Tensor(), "scores": Tensor(), "labels": IntTensor()}],
             [{"boxes": Tensor(), "labels": []}],
         )
+
+
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="test requires cuda")
+def test_device_changing():
+    """See issue: https://github.com/Lightning-AI/torchmetrics/issues/1743.
+
+    Checks that the custom apply function of the metric works as expected.
+    """
+    device = "cuda"
+    metric = MeanAveragePrecision(iou_type="segm").to(device)
+
+    for _ in range(2):
+        preds = []
+        targets = []
+        for _ in range(2):
+            result = {}
+            num_preds = torch.randint(0, 10, (1,)).item()
+            result["scores"] = torch.rand((num_preds,), device=device)
+            result["labels"] = torch.randint(0, 10, (num_preds,), device=device)
+            result["masks"] = torch.randint(0, 2, (num_preds, 10, 10), device=device).bool()
+            preds.append(result)
+            gt = {}
+            num_gt = torch.randint(0, 10, (1,)).item()
+            gt["labels"] = torch.randint(0, 10, (num_gt,), device=device)
+            gt["masks"] = torch.randint(0, 2, (num_gt, 10, 10), device=device).bool()
+            targets.append(gt)
```

Comment:
This is not part of the test case, just generating a sample, so let's move it outside...